File: DGV7.0-NON-SAAS/PayHub-Guest-Android/app/src/main/java/com/payhub/guest/api/GuestApiService.kt
package com.payhub.guest.api

import com.payhub.guest.data.model.AirtimeCatalogResponse
import com.payhub.guest.data.model.BettingCatalogResponse
import com.payhub.guest.data.model.CableCatalogResponse
import com.payhub.guest.data.model.CheckoutInitResponse
import com.payhub.guest.data.model.DataCatalogResponse
import com.payhub.guest.data.model.ElectricCatalogResponse
import com.payhub.guest.data.model.ExamCatalogResponse
import com.payhub.guest.data.model.GuestOrderStatusResponse
import com.payhub.guest.data.model.NetworkDetectResponse
import com.payhub.guest.data.model.VerifyCustomerResponse
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Query

/** Talks to the PHP endpoints under web/guest-api/ — the stateless, unauthenticated Guest Mode backend. */
interface GuestApiService {

    @GET("web/guest-api/catalog.php")
    suspend fun getAirtimeCatalog(@Query("service") service: String = "airtime"): Response<AirtimeCatalogResponse>

    @GET("web/guest-api/catalog.php")
    suspend fun getDataCatalog(@Query("service") service: String = "data"): Response<DataCatalogResponse>

    @GET("web/guest-api/catalog.php")
    suspend fun getCableCatalog(@Query("service") service: String = "cable"): Response<CableCatalogResponse>

    @GET("web/guest-api/catalog.php")
    suspend fun getElectricCatalog(@Query("service") service: String = "electric"): Response<ElectricCatalogResponse>

    @GET("web/guest-api/catalog.php")
    suspend fun getExamCatalog(@Query("service") service: String = "exam"): Response<ExamCatalogResponse>

    @GET("web/guest-api/catalog.php")
    suspend fun getBettingCatalog(@Query("service") service: String = "betting"): Response<BettingCatalogResponse>

    @GET("web/guest-api/identify-network.php")
    suspend fun identifyNetwork(@Query("phone") phone: String): Response<NetworkDetectResponse>

    @POST("web/guest-api/verify.php")
    suspend fun verifyCustomer(@Body body: Map<String, @JvmSuppressWildcards Any?>): Response<VerifyCustomerResponse>

    @POST("web/guest-api/checkout-init.php")
    suspend fun initCheckout(@Body body: Map<String, @JvmSuppressWildcards Any?>): Response<CheckoutInitResponse>

    @GET("web/guest-api/status.php")
    suspend fun getOrderStatus(@Query("reference") reference: String): Response<GuestOrderStatusResponse>
}

File: DGV7.0-NON-SAAS/PayHub-Guest-iOS/PayHubGuest/Networking/GuestAPIService.swift
import Foundation

/// Talks to web/guest-api/*.php — the stateless, unauthenticated Guest Mode backend.
/// Uses the platform default URLSession TLS trust (no custom trust-all delegate — see
/// this same security pass's fix to the Android RetrofitClient.kt for the bug this avoids).
final class GuestAPIService {

    static let shared = GuestAPIService()

    private let session: URLSession

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 30
        self.session = URLSession(configuration: config)
    }

    // MARK: - Catalog

    func getAirtimeCatalog() async -> ApiResult<AirtimeCatalogResponse> {
        await get(path: "web/guest-api/catalog.php", query: ["service": "airtime"])
    }
    func getDataCatalog() async -> ApiResult<DataCatalogResponse> {
        await get(path: "web/guest-api/catalog.php", query: ["service": "data"])
    }
    func getCableCatalog() async -> ApiResult<CableCatalogResponse> {
        await get(path: "web/guest-api/catalog.php", query: ["service": "cable"])
    }
    func getElectricCatalog() async -> ApiResult<ElectricCatalogResponse> {
        await get(path: "web/guest-api/catalog.php", query: ["service": "electric"])
    }
    func getExamCatalog() async -> ApiResult<ExamCatalogResponse> {
        await get(path: "web/guest-api/catalog.php", query: ["service": "exam"])
    }
    func getBettingCatalog() async -> ApiResult<BettingCatalogResponse> {
        await get(path: "web/guest-api/catalog.php", query: ["service": "betting"])
    }

    func identifyNetwork(phone: String) async -> ApiResult<NetworkDetectResponse> {
        await get(path: "web/guest-api/identify-network.php", query: ["phone": phone])
    }

    func verifyCustomer(body: [String: Any?]) async -> ApiResult<VerifyCustomerResponse> {
        await post(path: "web/guest-api/verify.php", body: body)
    }

    func initCheckout(body: [String: Any?]) async -> ApiResult<CheckoutInitResponse> {
        await post(path: "web/guest-api/checkout-init.php", body: body)
    }

    func getOrderStatus(reference: String) async -> ApiResult<GuestOrderStatusResponse> {
        await get(path: "web/guest-api/status.php", query: ["reference": reference])
    }

    // MARK: - Core request helpers

    private func get<T: Decodable>(path: String, query: [String: String]) async -> ApiResult<T> {
        guard var components = URLComponents(url: AppConfig.baseURL.appendingPathComponent(path), resolvingAgainstBaseURL: false) else {
            return .failure("Invalid request URL")
        }
        components.queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
        guard let url = components.url else { return .failure("Invalid request URL") }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("payhub-guest-ios", forHTTPHeaderField: "X-App-Source")

        return await execute(request)
    }

    private func post<T: Decodable>(path: String, body: [String: Any?]) async -> ApiResult<T> {
        let url = AppConfig.baseURL.appendingPathComponent(path)
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("payhub-guest-ios", forHTTPHeaderField: "X-App-Source")

        // Compact-map out nil values — JSONSerialization can't encode Optional<Any>.none directly.
        let cleaned = body.compactMapValues { $0 }
        guard let data = try? JSONSerialization.data(withJSONObject: cleaned) else {
            return .failure("Could not encode request")
        }
        request.httpBody = data

        return await execute(request)
    }

    private func execute<T: Decodable>(_ request: URLRequest) async -> ApiResult<T> {
        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                return .failure("No response from server")
            }
            let decoder = JSONDecoder()

            if (200...299).contains(http.statusCode) {
                if let decoded = try? decoder.decode(T.self, from: data) {
                    return .success(decoded)
                }
                return .failure("Server returned an unexpected response. Please try again.")
            } else {
                if let failure = try? decoder.decode(ApiFailure.self, from: data), let desc = failure.desc {
                    return .failure(desc)
                }
                return .failure("Request failed (\(http.statusCode))")
            }
        } catch {
            return .failure("Network error — please check your connection and try again.")
        }
    }
}
